src/modm/platform/dma/stm32/dma_hal.hpp.in

```diff
@@ -0,0 +1,304 @@
+	/// Register offset from channel to channel
+	static constexpr uint32_t CHANNEL_2_CHANNEL { 0x14 };
+
+	static void
+	clearInterruptFlags(uint32_t flags)
+	{
+		DMA_TypeDef *DMA = reinterpret_cast<DMA_TypeDef *>(DMA_BASE);
+		DMA->IFCR |= flags;
+	}
+
+	static uint32_t
+	getInterruptFlags()
+	{
+		DMA_TypeDef *DMA = reinterpret_cast<DMA_TypeDef *>(DMA_BASE);
+		return DMA->ISR;
+	}
+};
+
+/**
+ * Hardware abstraction layer of a DMA channel
+ *
+ * @tparam ChannelID the ID of the channel
+ * @tparam CHANNEL_BASE base address of the channel registers
+ *
+ * @author	Mike Wolfram
+ * @ingroup	modm_platform_dma
+ */
+template <DmaBase::Channel ChannelID, uint32_t CHANNEL_BASE>
+class DmaChannelHal : public DmaBase
+{
+	/**
+	 * Direction of DMA transfer
+	 *
+	 * Initialized to invalid value, must be configured before setting the
+	 * source and destination addresses.
+	 */
+	static inline DmaBase::DataTransferDirection direction {
+		modm::platform::DmaBase::DataTransferDirection(0xff)
+	};
+
+public:
+	/**
+	 * Configure the DMA channel (HAL)
+	 *
+	 * Stops the DMA channel and writes the new values to its control register.
+	 *
+	 * @param[in] dir Direction of the DMA channel
+	 * @param[in] memoryDataSize Size of data in memory (byte, halfword, word)
+	 * @param[in] peripheralDataSize Size of data in peripheral (byte, halfword, word)
+	 * @param[in] memoryIncrement Defines whether the memory address is incremented
+	 * 			  after a transfer completed
+	 * @param[in] peripheralIncrement Defines whether the peripheral address is
+	 * 			  incremented after a transfer completed
+	 * @param[in] priority Priority of the DMA channel
+	 * @param[in] circularMode Transfer data in circular mode?
+	 */
+	static void
+	configure(DataTransferDirection dir, MemoryDataSize memoryDataSize,
+			PeripheralDataSize peripheralDataSize,
+			MemoryIncrementMode memoryIncrement,
+			PeripheralIncrementMode peripheralIncrement, Priority priority,
+			CircularMode circular = CircularMode::Disabled)
+	{
+		stop();
+
+		DMA_Channel_TypeDef *Base = (DMA_Channel_TypeDef *) CHANNEL_BASE;
+		Base->CCR = uint32_t(dir) | uint32_t(memoryDataSize) |
+				uint32_t(peripheralDataSize) | uint32_t(memoryIncrement) |
+				uint32_t(peripheralIncrement) | uint32_t(priority) |
+				uint32_t(circular);
+		direction = dir;
+	}
+
+	/**
+	 * Start the transfer of the DMA channel
+	 */
+	static void
+	start();
+	/**
+	 * Stop a DMA channel transfer
+	 */
+	static void
+	stop();
+
+	/**
+	 * Get the direction of the data transfer
+	 */
+	static DataTransferDirection
+	getDataTransferDirection();
+
+	/**
+	 * Set the source and destination address of the DMA channel
+	 *
+	 * Requires to set the direction of the DMA channel first
+	 *
+	 * @param[in] sourceAddress Source address of DMA transfer
+	 * @param[in] destinationAddress Destination address of DMA transfer
+	 */
+	static void
+	setAddresses(uint32_t sourceAddress, uint32_t destinationAddress)
```

Comment:
```suggestion
	setAddresses(uint32_t* sourceAddress, uint32_t* destinationAddress)
```
or even
```suggestion
	setAddresses(void* sourceAddress, void* destinationAddress)
```

Reply:
What would you prefer @mikewolfram @chris-durand @salkinium?

Reply:
We could add some overloaded methods:
```cpp
setSourceAddress(uint8_t*);  // Implicitly sets the source data size to 8 bits
setSourceAddress(uint16_t*);  // Implicitly sets the source data size to 16 bits
setSourceAddress(uint32_t*);  // Implicitly sets the source data size to 32 bits
setSourceAddress(void*);  // Does not set the source data size
```
I'm not sure how evil that's gonna get. :joy:

Reply:
I totally overlooked the `uint32_t` there. Your suggestion `uint32_t*` doesn't feel right to me. The way to say "pointer to something unspecified which is not cv-qualified" is `void*`. With just one `void*` overload you would not get nice implicit conversion to `void*` in all cases because the source can be `const` and hardware registers are `volatile`. For that you need some overloads with permutations of `const` and `volatile` qualified `void* arguments`.

If we stayed with an int type for the address, it should be `intptr_t` or `uintptr_t` to express the intent.

Reply:
> We could add some overloaded methods:
> 
> ```c++
> setSourceAddress(uint8_t*);  // Implicitly sets the source data size to 8 bits
> setSourceAddress(uint16_t*);  // Implicitly sets the source data size to 16 bits
> setSourceAddress(uint32_t*);  // Implicitly sets the source data size to 32 bits
> setSourceAddress(void*);  // Does not set the source data size
> ```
> 
> I'm not sure how evil that's gonna get. joy

This feels a bit odd to me and easy to get wrong. A function overload set should have the same semantics for different overloads. In my opinion it is bad design, when the user has to know which one is chosen to understand the code, especially in times of `auto`.

Reply:
I don't like the idea to set the data size implicitly when setting source or destination addresses. Typically the size won't change, but the addresses are set many times. Also setting the size adds unnecessary overhead. When you look at the SPI example above you'll see that every call to _transfer()_ will set both the source and destination address.

I know in the original implementation there were dedicated functions to set the addresses for 8, 16 and 32 bit addresses. I think it's totally ok to have only one function. Changing the argument from uint32_t to uintptr_t makes sense.